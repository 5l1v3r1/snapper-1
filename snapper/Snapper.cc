/*
 * Copyright (c) 2011 Novell, Inc.
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact Novell, Inc.
 *
 * To contact Novell about this file by physical or electronic mail, you may
 * find current contact information at www.novell.com.
 */


#include <glob.h>
#include <map>
#include <iostream>

#include "snapper/Snapper.h"
#include "snapper/AppUtil.h"
#include "snapper/XmlFile.h"
#include "snapper/Enum.h"
#include "snapper/SnapperTmpl.h"


namespace snapper
{
    using namespace std;


    bool initialized = false;

    list<Snapshot> snapshots;


    Snapshot& snapshot1;
    Snapshot& snapshot2;


    std::ostream& operator<<(std::ostream& s, const Snapshot& x)
    {
	s << "type:" << toString(x.type) << " num:" << x.num;

	if (x.pre_num != 0)
	    s << " pre-num:" << x.pre_num;

	s << " date:" << x.date;

	if (!x.description.empty())
	    s << " description:" << x.description;

	return s;
    }


    bool operator<(Snapshot a, Snapshot b)
    {
	return a.num < b.num;
    }


    void
    readSnapshots()
    {
	list<string> infos = glob("/snapshots/*/snapshot.info", GLOB_NOSORT);
	for (list<string>::const_iterator it = infos.begin(); it != infos.end(); ++it)
	{
	    unsigned int num;
	    it->substr(11) >> num;

	    XmlFile file(*it);
	    const xmlNode* root = file.getRootElement();
	    const xmlNode* node = getChildNode(root, "snapshot");

	    Snapshot snapshot;

	    string tmp;

	    if (getChildValue(node, "type", tmp))
	    {
		if (!toValue(tmp, snapshot.type, true))
		{
		}
	    }

	    getChildValue(node, "num", snapshot.num);
	    assert(num == snapshot.num);

	    getChildValue(node, "date", snapshot.date);

	    getChildValue(node, "description", snapshot.description);

	    getChildValue(node, "pre_num", snapshot.pre_num);

	    snapshots.push_back(snapshot);
	}

	snapshots.sort();
    }


    void
    initialize()
    {
	initialized = true;

	readSnapshots();
    }


    void
    assertInit()
    {
	if (!initialized)
	    initialize();
    }


    bool
    getSnapshot(unsigned int num, Snapshot& snapshot)
    {
	assertInit();

	for (list<Snapshot>::const_iterator it = snapshots.begin();
	     it != snapshots.end(); ++it)
	{
	    if (it->num == num)
	    {
		snapshot = *it;
		return true;
	    }
	}

	return false;
    }


    const list<Snapshot>&
    getSnapshots()
    {
	assertInit();

	return snapshots;
    }


    void
    listSnapshots()
    {
	assertInit();

	for (list<Snapshot>::const_iterator it = snapshots.begin();
	     it != snapshots.end(); ++it)
	{
	    cout << *it << endl;
	}
    }


    unsigned int
    nextSnapshotNumber()
    {
	assertInit();

	unsigned int num = 1;

	if (!snapshots.empty())
	    num = snapshots.rbegin()->num + 1;

	return num;
    }


    bool
    writeInfo(const Snapshot& snapshot)
    {
	createPath("/snapshots/" + decString(snapshot.num));

	XmlFile xml;
	xmlNode* node = xmlNewNode("snapshot");
	xml.setRootElement(node);

	setChildValue(node, "type", toString(snapshot.type));

	setChildValue(node, "num", snapshot.num);

	setChildValue(node, "date", snapshot.date);

	if (snapshot.type == SINGLE || snapshot.type == PRE)
	    setChildValue(node, "description", snapshot.description);

	if (snapshot.type == POST)
	    setChildValue(node, "pre_num", snapshot.pre_num);

	xml.save("/snapshots/" + decString(snapshot.num) + "/snapshot.info");

	return true;
    }


    unsigned int
    createSingleSnapshot(string description)
    {
	Snapshot snapshot;
	snapshot.type = SINGLE;
	snapshot.num = nextSnapshotNumber();
	snapshot.date = datetime();
	snapshot.description = description;

	snapshots.push_back(snapshot);
	writeInfo(snapshot);

	return snapshot.num;
    }


    unsigned int
    createPreSnapshot(string description)
    {
	Snapshot snapshot;
	snapshot.type = PRE;
	snapshot.num = nextSnapshotNumber();
	snapshot.date = datetime();
	snapshot.description = description;

	snapshots.push_back(snapshot);
	writeInfo(snapshot);

	return snapshot.num;
    }


    unsigned int
    createPostSnapshot(unsigned int pre_num)
    {
	Snapshot snapshot;
	snapshot.type = POST;
	snapshot.num = nextSnapshotNumber();
	snapshot.date = datetime();
	snapshot.pre_num = pre_num;

	snapshots.push_back(snapshot);
	writeInfo(snapshot);

	return snapshot.num;
    }



    bool
    setComparisonNums(unsigned int num1, unsigned int num2)
    {
	if (num1 == 0 || !getSnapshot(num1, snapshot1))
	    return false;

	if (num2 != 0 && !getSnapshot(num2, snapshot2))
	    return false;

	if (snapshot1.num != snapshot2.pre_num)
	    return false;

	// load or generate file list

	return true;
    }

}
