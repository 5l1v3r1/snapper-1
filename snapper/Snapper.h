/*
 * Copyright (c) 2011 Novell, Inc.
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact Novell, Inc.
 *
 * To contact Novell about this file by physical or electronic mail, you may
 * find current contact information at www.novell.com.
 */


#ifndef SNAPPER_H
#define SNAPPER_H


#include "snapper/Snapshot.h"
#include "snapper/File.h"


namespace snapper
{
    class SysconfigFile;


    struct CompareCallback
    {
	CompareCallback() {}
	virtual ~CompareCallback() {}

	virtual void start() {}
	virtual void stop() {}
    };


    class Snapper
    {
    public:

	Snapper(const string& config_name = "root");
	~Snapper();

	string subvolumeDir() const;
	string snapshotsDir() const;

	Snapshots& getSnapshots() { return snapshots; }
	const Snapshots& getSnapshots() const { return snapshots; }

	Snapshots::const_iterator getSnapshotCurrent() const;

	Snapshots::iterator createSingleSnapshot(string description);
	Snapshots::iterator createPreSnapshot(string description);
	Snapshots::iterator createPostSnapshot(Snapshots::const_iterator pre);

	void deleteSnapshot(Snapshots::iterator snapshot);

	void startBackgroundComparsion(Snapshots::const_iterator snapshot1,
				       Snapshots::const_iterator snapshot2);

	bool setComparison(Snapshots::const_iterator snapshot1,
			   Snapshots::const_iterator snapshot2);

	Snapshots::const_iterator getSnapshot1() const { return snapshot1; }
	Snapshots::const_iterator getSnapshot2() const { return snapshot2; }

	Files& getFiles() { return files; }
	const Files& getFiles() const { return files; }

	RollbackStatistic getRollbackStatistic() const;

	bool doRollback();

	bool doCleanupNumber();
	bool doCleanupTimeline();

	void setCompareCallback(CompareCallback* p) { compare_callback = p; }
	CompareCallback* getCompareCallback() const { return compare_callback; }

    private:

	void filter1(list<Snapshots::iterator>& tmp1);

	const string config_name;

	SysconfigFile* config;

	string subvolume;

	Snapshots snapshots;

	Snapshots::const_iterator snapshot1;
	Snapshots::const_iterator snapshot2;

	Files files;

	CompareCallback* compare_callback;

    };

};


#endif
