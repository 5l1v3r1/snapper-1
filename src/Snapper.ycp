/* ------------------------------------------------------------------------------
 * Copyright (c) 2006 Novell, Inc. All Rights Reserved.
 *
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of version 2 of the GNU General Public License as published by the
 * Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, contact Novell, Inc.
 *
 * To contact Novell about this file by physical or electronic mail, you may find
 * current contact information at www.novell.com.
 * ------------------------------------------------------------------------------
 */

/**
 * File:	modules/Snapper.ycp
 * Summary:	Snapper settings, input and output functions
 * Authors:	Jiri Suchomel <jsuchome@suse.cz>
 *
 * $Id$
 *
 * Representation of the configuration of snapper.
 * Input and output routines.
 */

{

module "Snapper";
textdomain "snapper";

import "FileUtils";
import "Label";
import "Progress";


// global list of all snapshot
global list<map> snapshots	= [];

global string snapshots_path	= "/snapshots"; // FIXME

global map selected_snapshot	= $[];

// mapping of snapshot number to index in snapshots list
global map<integer,integer> id2index	= $[];

// index to snapshots list
global integer selected_snapshot_index	= 0;

/**
 * Return list of files modified between given snapshots
 * File is a map with name (path) and some parameters
 */
global map<string, map> ReadModifiedFiles (integer from, integer to) {
    return (map<string, map>) SCR::Read (.snapper.diff_tree, $[ "from" : from, "to" : to]);
}

/**
 * Describe what was done with given file between given snapshots
 * - when new is 0, meaning is 'current system'
 */
global map GetFileModification (string file, integer old, integer new) {

    map ret	= $[];

    string file1	= sformat ("%1/%2/snapshot%3", snapshots_path, old, file);
    string file2	= sformat ("%1/%2/snapshot%3", snapshots_path, new, file);
    if (new == 0)
	file2	= file;

    y2milestone ("comparing '%1' and '%2'", file1, file2);

    if (FileUtils::Exists (file1) && FileUtils::Exists (file2))
    {
	list<string> status	= [ "no_change" ];
	map out	= (map) SCR::Execute (.target.bash_output,
	    sformat ("/usr/bin/diff -u %1 %2", file1, file2));
	if (out["stderr"]:"" != "")
	{
	    y2warning ("out: %1", out);
	    ret["diff"] = out["stderr"]:"";
	}
	// the file diff
	else if (out["stdout"]:nil != "")
	{
	    status	= ["diff"];
	    ret["diff"]	= out["stdout"]:"";
	}

	// check mode and ownerships
	out = (map) SCR::Execute (.target.bash_output,
	    sformat ("ls -l %1 %2 | cut -f 1,3,4 -d ' '", file1, file2));
	list<string> parts	= splitstring (out["stdout"]:""," \n");

	if (parts[0]:"" != parts[3]:"")
	{
	    status		= add (status, "mode");
	    ret["mode1"]	= parts[0]:"";
	    ret["mode2"]	= parts[3]:"";
	}
	if (parts[1]:"" != parts[4]:"")
	{
	    status		= add (status, "user");
	    ret["user1"]	= parts[1]:"";
	    ret["user2"]	= parts[4]:"";
	}
	if (parts[2]:"" != parts[5]:"")
	{
	    status		= add (status, "group");
	    ret["group1"]	= parts[2]:"";
	    ret["group2"]	= parts[5]:"";
	}
	ret["status"]	= status;
    }
    else if (FileUtils::Exists (file1))
    {
	ret["status"]	= ["removed"];
    }
    else if (FileUtils::Exists (file2))
    {
	ret["status"]	= ["created"];
    }
    else
    {
	ret["status"]	= ["none"];
    }
    return ret;
}

/**
 * Read the list of snapshots
 */
global boolean ReadSnapshots () {


    list<map> snapshot_maps	= (list<map>) SCR::Read (.snapper.snapshots);
    if (snapshot_maps == nil) snapshot_maps = [];
    integer i	= 0;
    foreach (map snapshot, snapshot_maps, {
	integer id	= snapshot["num"]:0;
	if (id == 0) return; // ignore the 'current system'
	snapshot["name"]	= tostring (id);
	y2debug ("snapshot data: %1", snapshot);
	snapshots	= add (snapshots, snapshot);
	id2index[id]	= i;
	i		= i + 1;
    });


    return true;
}

/**
 * Read all snapper settings
 * @return true on success
 */
global boolean Read() {

    /* Snapper read dialog caption */
    string caption = _("Initializing Snapper");

    integer steps = 2;

    // We do not set help text here, because it was set outside
    Progress::New( caption, " ", steps, [
	    /* Progress stage 1/3 */
	    _("Read the list of snapshots"),
	], [
	    /* Progress step 1/3 */
	    _("Reading the database..."),
	    /* Progress finished */
	    _("Finished")
	],
	""
    );

    Progress::NextStage();
    // read database

    ReadSnapshots ();

    Progress::NextStage();
    return true;
}

/**
 * Copy given files from selected snapshot to current filesystem
 */
global boolean RestoreFiles (integer snapshot_num, list<string> files) {

    boolean ret	= true;
    y2milestone ("going to restore files %1", files);

    UI::OpenDialog (`opt (`decorated), `HBox (`HSpacing(1.5), `VBox (
	`HSpacing (60),
	// label for log window
	`LogView (`id (`log), _("Restoring Files..."), 8, 0),
	`ProgressBar (`id (`progress), "", size (files), 0),
	`PushButton (`id (`ok), Label::OKButton ())
    ), `HSpacing (1.5)));

    UI::ChangeWidget (`id (`ok), `Enabled, false);
    integer progress	= 0;
    foreach (string file, files, {

	UI::ChangeWidget (`id (`progress ), `Value, progress);

	string orig	= sformat ("%1/%2/snapshot%3", snapshots_path, snapshot_num, file);

	string dir	= substring (file, 0, findlastof (file, "/"));
	y2milestone ("copying file %1 to %2 (dir: %3)", orig, file, dir);
	if (FileUtils::CheckAndCreatePath (dir))
	{
	    SCR::Execute (.target.bash, sformat ("/bin/cp -a %1 %2", orig, file));
	    UI::ChangeWidget (`id (`log), `LastLine, file + "\n");
	}
	else
	{
	    // log entry (%1 is file name)
	    UI::ChangeWidget (`id (`log), `LastLine, sformat (_("%1 skipped\n"), file));
	}
	sleep (100);
	progress        = progress + 1;
    });

    UI::ChangeWidget (`id (`progress ), `Value, progress);
    UI::ChangeWidget (`id (`ok), `Enabled, true);

    UI::UserInput ();
    UI::CloseDialog ();

    return ret;
}


/* EOF */
}
