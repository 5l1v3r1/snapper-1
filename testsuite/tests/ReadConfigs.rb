# encoding: utf-8

module Yast
  class ReadConfigsClient < Client
    def main
      # testedfiles: Snapper.ycp

      Yast.import "Testsuite"
      Yast.import "Snapper"

      @READ = { "snapper" => { "configs" => [] } }

      Testsuite.Test(lambda { Snapper.ReadConfigs }, [@READ, {}, {}], 0)
      Testsuite.Test(lambda { Snapper.current_config }, [@READ, {}, {}], 0)

      Ops.set(@READ, ["snapper", "configs"], nil)

      Testsuite.Test(lambda { Snapper.ReadConfigs }, [@READ, {}, {}], 0)
      Testsuite.Test(lambda { Snapper.current_config }, [@READ, {}, {}], 0)

      Ops.set(@READ, ["snapper", "configs"], ["opt", "var", "root"])

      Testsuite.Test(lambda { Snapper.ReadConfigs }, [@READ, {}, {}], 0)
      Testsuite.Test(lambda { Snapper.current_config }, [@READ, {}, {}], 0)

      Ops.set(@READ, ["snapper", "configs"], ["opt", "var"])

      Testsuite.Test(lambda { Snapper.ReadConfigs }, [@READ, {}, {}], 0)
      Testsuite.Test(lambda { Snapper.current_config }, [@READ, {}, {}], 0)

      nil
    end
  end
end

Yast::ReadConfigsClient.new.main
