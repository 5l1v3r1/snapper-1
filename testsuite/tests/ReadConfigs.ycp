{
    // testedfiles: Snapper.ycp

    import "Testsuite";
    import "Snapper";

    map READ	= $[
	"snapper"	: $[
	    "configs"	: []
	]
    ];

    Testsuite::Test (``(Snapper::ReadConfigs ()), [ READ, $[], $[] ], 0);
    Testsuite::Test (``(Snapper::current_config), [ READ, $[], $[] ], 0);

    READ["snapper","configs"]	= nil;

    Testsuite::Test (``(Snapper::ReadConfigs ()), [ READ, $[], $[] ], 0);
    Testsuite::Test (``(Snapper::current_config), [ READ, $[], $[] ], 0);

    READ["snapper","configs"]	= [ "opt", "var", "root" ];

    Testsuite::Test (``(Snapper::ReadConfigs ()), [ READ, $[], $[] ], 0);
    Testsuite::Test (``(Snapper::current_config), [ READ, $[], $[] ], 0);

    READ["snapper","configs"]	= [ "opt", "var" ];

    Testsuite::Test (``(Snapper::ReadConfigs ()), [ READ, $[], $[] ], 0);
    Testsuite::Test (``(Snapper::current_config), [ READ, $[], $[] ], 0);

}
