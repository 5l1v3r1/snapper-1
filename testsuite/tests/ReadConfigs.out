Read	.snapper.configs []
Return	[]
Return	root
Read	.snapper.configs nil
Return	["root"]
Return	root
Read	.snapper.configs ["opt", "var", "root"]
Return	["opt", "var", "root"]
Return	root
Read	.snapper.configs ["opt", "var"]
Return	["opt", "var"]
Return	opt
